### PR TITLE
REST and CLI to get function metrics in json for monitoring

### PR DESCRIPTION
**Pull request by rdhabalia**
### Motivation

right now, function metrics api returns metrics in prometheus format which is not user-friendly and hard to parse. So, we need metrics in json format and cli to fetch metrics same as what we have for broker.
existing metrics:
```
pulsar_function__total_processed__count{cluster="default", namespace="sample/ns1", name="demo1", instanceId="0"} 0.0 1533172228038
pulsar_function__total_processed__max{cluster="default", namespace="sample/ns1", name="demo1", instanceId="0"} 0.0 1533172228039
pulsar_function__total_processed__min{cluster="default", namespace="sample/ns1", name="demo1", instanceId="0"} 0.0 1533172228039
pulsar_function__total_processed__sum{cluster="default", namespace="sample/ns1", name="demo1", instanceId="0"} 0.0 1533172228039
pulsar_function__total_successfully_processed__count{cluster="default", namespace="sample/ns1", name="demo1", instanceId="0"} 0.0 1533172228040
pulsar_function__total_successfully_processed__max{cluster="default", namespace="sample/ns1", name="demo1", instanceId="0"} 0.0 1533172228040
pulsar_function__total_successfully_processed__min{cluster="default", namespace="sample/ns1", name="demo1", instanceId="0"} 0.0 1533172228040
pulsar_function__total_successfully_processed__sum{cluster="default", namespace="sample/ns1", name="demo1", instanceId="0"} 0.0 1533172228040
pulsar_function__total_system_exceptions__count{cluster="default", namespace="sample/ns1", name="demo1", instanceId="0"} 0.0 1533172228040
pulsar_function__total_system_exceptions__max{cluster="default", namespace="sample/ns1", name="demo1", instanceId="0"} 0.0 1533172228041
```

json metrics:
```
./pulsar-admin  functions metrics
{
  "metrics": [
    {
      "name": "sample/ns1/demo3",
      "metricsData": {
        "metrics": {
          "__total_processed__": {},
          "__total_successfully_processed__": {},
          "__total_system_exceptions__": {},
          "__total_user_exceptions__": {},
          "__total_serialization_exceptions__": {},
          "__avg_latency_ms__": {}
        }
      }
    },
    {
      "name": "sample/ns1/demo1",
      "metricsData": {
        "metrics": {
          "__total_processed__": {},
          "__total_successfully_processed__": {},
          "__total_system_exceptions__": {},
          "__total_user_exceptions__": {},
          "__total_serialization_exceptions__": {},
          "__avg_latency_ms__": {}
        }
      }
    }
  ]
}

```

### Modifications

- add REST api for metrics
- add cli to get metrics in json format

### CODE REVIEW

**Comment by rdhabalia** [comment]
rerun cpp tests

**Comment by rdhabalia** [comment]
rerun integration tests

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/Functions.java [hunk 1]
```diff
@@ -259,4 +260,11 @@
      * @throws PulsarAdminException 
      */
     List<WorkerInfo> getCluster() throws PulsarAdminException;
+    
+    /**
+     * Get function worker metrics
+     * @return
+     * @throws PulsarAdminException 
+     */
+    Metrics getMetrics() throws PulsarAdminException;
```
Comment (srkukarni) [inline pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/Functions.java]: Then maybe we change this to getWorkerMetrics?

Comment (srkukarni) [reply pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/Functions.java]: I'm still a bit confused here. Are you getting worker related metrics or function related metrics?

Comment (rdhabalia) [reply pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/Functions.java]: same metrics which is being exposed on `/metrics` endpoint for prometheus but it will be in json format. it will be used by our non-prometheus monitoring system to aggregate all functions-metrics owned  by the worker.

Comment (srkukarni) [reply pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/Functions.java]: I see a disconnect here. Most of other calls listed here are all function specific that get routed appropriately to the right worker who then answers with the right information.
On the other hand, this particular call is for getting the metrics housed by a specific worker. Do we want a different place to house such calls?

Comment (rdhabalia) [reply pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/Functions.java]: > On the other hand, this particular call is for getting the metrics housed by a specific worker. Do we want a different place to house such calls?

hmm.. yes. right now, in broker we have separate resource [BrokerStats](https://github.com/apache/incubator-pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/BrokerStatsBase.java) to serve broker-localhost stats, probably we can add resource: `WorkerStats` for consistency and we can add worker-specific stats in the same resource as needed.

Comment (rdhabalia) [reply pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/Functions.java]: changed it and added separate endpoint.

**Comment by rdhabalia** [comment]
@srkukarni @sijie @jerrypeng @merlimat @jai1 
addressed the change regarding endpoint. can you please give another look.

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/FunctionsImpl.java [hunk 2]
```diff
@@ -47,6 +47,7 @@
 import org.apache.pulsar.common.policies.data.ErrorData;
 import org.apache.pulsar.functions.proto.Function.FunctionDetails;
 import org.apache.pulsar.functions.proto.InstanceCommunication.FunctionStatusList;
+import org.apache.pulsar.functions.proto.InstanceCommunication.Metrics;
```
Comment (srkukarni) [inline pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/FunctionsImpl.java]: leftover stuff?

**Comment by srkukarni** [comment]
rerun integration tests
